### PR TITLE
Implement local auth for login and signup

### PR DESCRIPTION
**Pull request by PrinceOm**

### CODE REVIEW

File: api/user/user.model.js [hunk 1]
```diff
@@ -1,8 +1,28 @@
 const mongoose = require('mongoose');
+const bcrypt = require('bcrypt-nodejs');
+const Promise = require('bluebird');
 
-const userSchema = new mongoose.Schema({
+
```
Comment (dting) [inline api/user/user.model.js]: Extra line here.

File: api/user/user.model.js [hunk 2]
```diff
@@ -1,8 +1,28 @@
...
+const userSchema = mongoose.Schema({
```
Comment (dting) [inline api/user/user.model.js]: Does this pass the linter? I think we want to use keyword new to construct the new Schema.

File: api/user/index.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+const express = require('express');
+
+const router = express.Router();
+const { login, signUp } = require('./user.controller');
```
Comment (dting) [inline api/user/index.js]: Let's move this to the top next to the other require statements.

File: config/environment/index.js [hunk 4]
```diff
@@ -4,6 +4,14 @@ const baseSettings = {
   env: process.env.NODE_ENV || 'development',
   port: process.env.PORT || 8888,
   ip: process.env.IP || '0.0.0.0',
+  secrets: {
+    session: process.env.SESSION_SECRET || 'session-secret',
+  },
+  facebook: {
```
Comment (dting) [inline config/environment/index.js]: Let's remove this facebook section for now (until we start adding the facebook oauth stuff.

File: api/user/user.model.js [hunk 5]
```diff
@@ -1,8 +1,28 @@
...
   username: { type: String, required: true, index: { unique: true } },
   password: { type: String, required: true },
 });
 
-module.exports = mongoose.model('User', userSchema);
+const User = mongoose.model('User', userSchema);
+
+User.comparePassword = (candidatePassword, savedPassword) => {
+  const compare = Promise.promisify(bcrypt.compare);
```
Comment (dting) [inline api/user/user.model.js]: I'm wondering if these's promisification of the helper functions should be outside the function bodies. That way they aren't promisified every time the function is called.

File: api/user/user.model.js [hunk 6]
```diff
@@ -1,8 +1,28 @@
...
+  return compare(candidatePassword, savedPassword)
+    .then(result => result);
+};
+
+userSchema.pre('save', function (next) {
+  const cipher = Promise.promisify(bcrypt.hash);
```
Comment (dting) [inline api/user/user.model.js]: Same as above.

File: api/user/user.model.js [hunk 7]
```diff
@@ -1,8 +1,28 @@
...
   password: { type: String, required: true },
 });
 
-module.exports = mongoose.model('User', userSchema);
+const User = mongoose.model('User', userSchema);
+
+User.comparePassword = (candidatePassword, savedPassword) => {
+  const compare = Promise.promisify(bcrypt.compare);
+  return compare(candidatePassword, savedPassword)
+    .then(result => result);
```
Comment (dting) [inline api/user/user.model.js]: Do we need this noop `then` function?

**Comment by dting** [comment]
Looks really good OM. Had some comments, let me know when to take another look.

File: api/user/user.model.js (hunk 1 above)
Comment (PrinceOm) [inline api/user/user.model.js]: done

File: api/user/user.model.js (hunk 2 above)
Comment (PrinceOm) [inline api/user/user.model.js]: Done.  It does pass with 'new'.

File: api/user/index.js (hunk 3 above)
Comment (PrinceOm) [inline api/user/index.js]: Done

File: config/environment/index.js (hunk 4 above)
Comment (PrinceOm) [inline config/environment/index.js]: Done.

File: api/user/user.model.js (hunk 5 above)
Comment (PrinceOm) [inline api/user/user.model.js]: Done.

File: api/user/user.model.js (hunk 6 above)
Comment (PrinceOm) [inline api/user/user.model.js]: Done.

File: api/user/user.model.js (hunk 7 above)
Comment (PrinceOm) [inline api/user/user.model.js]: Done.

File: package.json [hunk 8]
```diff
@@ -41,12 +41,16 @@
     "supertest-as-promised": "^3.2.0"
   },
   "dependencies": {
+    "bcrypt-nodejs": "0.0.3",
     "bluebird": "^3.4.1",
     "body-parser": "^1.15.2",
+    "composable-middleware": "^0.3.0",
```
Comment (dting) [inline package.json]: This doesn't seem used yet. Let's remove it until we use it.

**Comment by dting** [comment]
LGTM. Just remove the unused package and ship it. Thanks.

**Pull request merged**